### PR TITLE
feat: module account address derivation

### PR DESCRIPTION
**Pull request by damiannolan**
## Description

- Adding module account to `interchain-accounts` for address derivation

closes: #303 

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#pr-targeting))
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [x] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [x] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#testing)
- [x] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [x] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [x] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Review `Codecov Report` in the comment section below once CI passes

### CODE REVIEW

File: modules/apps/27-interchain-accounts/keeper/keeper_test.go [hunk 1]
```diff
@@ -4,21 +4,26 @@ import (
...
 
 	"github.com/cosmos/ibc-go/modules/apps/27-interchain-accounts/types"
 	channeltypes "github.com/cosmos/ibc-go/modules/core/04-channel/types"
 	ibctesting "github.com/cosmos/ibc-go/testing"
 )
 
 var (
+	// TestAccAddress defines a resuable bech32 address for testing purposes
+	// TODO: update crypto.AddressHash() when sdk uses address.Module()
+	TestAccAddress = types.GenerateAddress(sdk.AccAddress(crypto.AddressHash([]byte(types.ModuleName))), TestPortID)
```
Comment (damiannolan) [inline modules/apps/27-interchain-accounts/keeper/keeper_test.go:20]: Presumably the sdk will eventually be using `address.Module()` from ADR-28 here: https://github.com/cosmos/cosmos-sdk/blob/master/x/auth/types/account.go#L164

So we'll need to replace `crypto.AddressHash()` when that's added

File: modules/apps/27-interchain-accounts/keeper/keeper.go [hunk 2]
```diff
@@ -163,3 +171,22 @@ func (k Keeper) SetInterchainAccountAddress(ctx sdk.Context, portID string, addr
 	store := ctx.KVStore(k.storeKey)
 	store.Set(types.KeyOwnerAccount(portID), []byte(address))
 }
+
+// NegotiateAppVersion handles application version negotation for the IBC interchain accounts module
+func (k Keeper) NegotiateAppVersion(
```
Comment (damiannolan) [inline modules/apps/27-interchain-accounts/keeper/keeper.go:176]: Moved the implementation of this from `AppModule` to the Keeper in order to gain access to the `accountKeeper` without having to expose it.
AppModule now just invokes this func on the Keeper.

File: modules/apps/27-interchain-accounts/keeper/keeper.go [hunk 3]
```diff
@@ -37,6 +39,12 @@ func NewKeeper(
 	channelKeeper types.ChannelKeeper, portKeeper types.PortKeeper,
 	accountKeeper types.AccountKeeper, scopedKeeper capabilitykeeper.ScopedKeeper, msgRouter *baseapp.MsgServiceRouter, hook types.IBCAccountHooks,
 ) Keeper {
+
+	// ensure ibc interchain accounts module account is set
+	if addr := accountKeeper.GetModuleAddress(types.ModuleName); addr == nil {
+		panic("the IBC interchain accounts module account has not been set")
```
Comment (seantking) [inline modules/apps/27-interchain-accounts/keeper/keeper.go:45]:
```suggestion
		panic("the Interchain Accounts module account has not been set")
```

**Review by seantking — APPROVED** [review]

File: modules/apps/27-interchain-accounts/types/account.go [hunk 4]
```diff
@@ -7,17 +7,17 @@ import (
...
-	"github.com/tendermint/tendermint/crypto/tmhash"
 	yaml "gopkg.in/yaml.v2"
 
 	connectiontypes "github.com/cosmos/ibc-go/v2/modules/core/03-connection/types"
 )
 
-// GenerateAddress returns an sdk.AccAddress using the provided port identifier
-func GenerateAddress(portID string) sdk.AccAddress {
-	return sdk.AccAddress(tmhash.SumTruncated([]byte(portID)))
+// GenerateAddress returns an sdk.AccAddress derived using the provided module account address and port identifier
```
Comment (colin-axner) [inline modules/apps/27-interchain-accounts/types/account.go:18]: Maybe we should mention this is a sub account of the module account?

File: modules/apps/27-interchain-accounts/types/account.go [hunk 5]
```diff
@@ -7,17 +7,17 @@ import (
 
 	crypto "github.com/cosmos/cosmos-sdk/crypto/types"
 	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/types/address"
```
Comment (colin-axner) [inline modules/apps/27-interchain-accounts/types/account.go:10]: nit: alias `sdkaddress`

File: modules/apps/27-interchain-accounts/types/account.go [hunk 6]
```diff
@@ -7,17 +7,17 @@ import (
...
 
 	connectiontypes "github.com/cosmos/ibc-go/v2/modules/core/03-connection/types"
 )
 
-// GenerateAddress returns an sdk.AccAddress using the provided port identifier
-func GenerateAddress(portID string) sdk.AccAddress {
-	return sdk.AccAddress(tmhash.SumTruncated([]byte(portID)))
+// GenerateAddress returns an sdk.AccAddress derived using the provided module account address and port identifier
+func GenerateAddress(moduleAccAddr sdk.AccAddress, portID string) sdk.AccAddress {
+	return sdk.AccAddress(address.Derive(moduleAccAddr, []byte(portID)))
```
Comment (colin-axner) [inline modules/apps/27-interchain-accounts/types/account.go:20]: Maybe we should specify that the portID must be the controller port ID

We should also double check this portID will always be unique (but I'm fairly certain that's already a requirement for ICA security)

File: modules/apps/27-interchain-accounts/keeper/keeper_test.go (hunk 1 above)
Comment (colin-axner) [reply modules/apps/27-interchain-accounts/keeper/keeper_test.go:20]: Could we open an issue to track this? Should we open an issue on the SDK?

**Review by colin-axner — APPROVED** [review]
Nice!

File: modules/apps/27-interchain-accounts/keeper/keeper_test.go (hunk 1 above)
Comment (damiannolan) [reply modules/apps/27-interchain-accounts/keeper/keeper_test.go:20]: Sure, I can open an issue on the SDK for this. I'll cc you

File: modules/apps/27-interchain-accounts/types/account.go (hunk 5 above)
Comment (damiannolan) [reply modules/apps/27-interchain-accounts/types/account.go]: Updated! :)

File: modules/apps/27-interchain-accounts/types/account.go (hunk 4 above)
Comment (damiannolan) [reply modules/apps/27-interchain-accounts/types/account.go]: Updated to be more explicit

**Pull request merged**